### PR TITLE
Do not remove closures during `ReplaceBodyWithLoop` pass

### LINKED ISSUE #71820: Invalid lowering of macro in closure
When trying to fix https://github.com/rust-lang/rust/issues/71104 I ended up fighting with some rustdoc test about declarative macro: https://github.com/rust-lang/rust/blob/master/src/test/rustdoc/macro-in-closure.rs

The problem is that the macro is properly lowered to hir because its exported but the parent item is not:
* Macro has def id `DefId(0:5 ~ macro_in_closure[317d]::main[0]::{{closure}}[0]::m[0])`
* Closure has def id `DefId(0:4 ~ macro_in_closure[317d]::main[0]::{{closure}}[0])`

The latter appears nowhere in the hir tree and when https://github.com/rust-lang/rust/blob/master/src/librustc_privacy/lib.rs#L945 tries to access its parent, it returns a `DefId` that has no associated `HirId`.

I don't know exactly how lowering of closure works, neither the details of declarative macro visibility, but i think the bug is either:
* the closure should have been lowered somehow
* the declarative macro should not have been lowered as it is probably not visible outside of the closure

Here is a hir dump of the test case:
```
warning: unused macro definition
 --> src/test/rustdoc/macro-in-closure.rs:7:9
  |
7 |         macro m() {}
  |         ^^^^^^^^^^^^
  |
  = note: `#[warn(unused_macros)]` on by default

Crate {
    item: CrateItem {
        module: Mod {
            inner: src/test/rustdoc/macro-in-closure.rs:3:1: 9:2,
            item_ids: [
                ItemId {
                    id: HirId {
                        owner: DefId(0:1 ~ macro_in_closure[317d]::{{misc}}[0]),
                        local_id: 0,
                    },
                },
                ItemId {
                    id: HirId {
                        owner: DefId(0:2 ~ macro_in_closure[317d]::std[0]),
                        local_id: 0,
                    },
                },
                ItemId {
                    id: HirId {
                        owner: DefId(0:3 ~ macro_in_closure[317d]::main[0]),
                        local_id: 0,
                    },
                },
            ],
        },
        attrs: [
            Attribute {
                kind: Normal(
                    AttrItem {
                        path: Path {
                            span: src/test/rustdoc/macro-in-closure.rs:3:4: 3:11,
                            segments: [
                                PathSegment {
                                    ident: feature#0,
                                    id: NodeId(2),
                                    args: None,
                                },
                            ],
                        },
                        args: Delimited(
                            DelimSpan {
                                open: src/test/rustdoc/macro-in-closure.rs:3:11: 3:12,
                                close: src/test/rustdoc/macro-in-closure.rs:3:22: 3:23,
                            },
                            Parenthesis,
                            TokenStream(
                                [
                                    (
                                        Token(
                                            Token {
                                                kind: Ident(
                                                    "decl_macro",
                                                    false,
                                                ),
                                                span: src/test/rustdoc/macro-in-closure.rs:3:12: 3:22,
                                            },
                                        ),
                                        NonJoint,
                                    ),
                                ],
                            ),
                        ),
                    },
                ),
                id: AttrId(0),
                style: Inner,
                span: src/test/rustdoc/macro-in-closure.rs:3:1: 3:24,
            },
        ],
        span: src/test/rustdoc/macro-in-closure.rs:3:1: 9:2,
    },
    exported_macros: [
        MacroDef {
            ident: m#0,
            vis: Spanned {
                node: Inherited,
                span: src/test/rustdoc/macro-in-closure.rs:7:9: 7:9,
            },
            attrs: [],
            hir_id: HirId {
                owner: DefId(0:5 ~ macro_in_closure[317d]::main[0]::{{closure}}[0]::m[0]),
                local_id: 0,
            },
            span: src/test/rustdoc/macro-in-closure.rs:7:9: 7:21,
            ast: MacroDef {
                body: Delimited(
                    DelimSpan {
                        open: src/test/rustdoc/macro-in-closure.rs:7:16: 7:16,
                        close: src/test/rustdoc/macro-in-closure.rs:7:21: 7:21,
                    },
                    Brace,
                    TokenStream(
                        [
                            (
                                Delimited(
                                    DelimSpan {
                                        open: src/test/rustdoc/macro-in-closure.rs:7:16: 7:17,
                                        close: src/test/rustdoc/macro-in-closure.rs:7:17: 7:18,
                                    },
                                    Paren,
                                    TokenStream(
                                        [],
                                    ),
                                ),
                                NonJoint,
                            ),
                            (
                                Token(
                                    Token {
                                        kind: FatArrow,
                                        span: src/test/rustdoc/macro-in-closure.rs:7:18: 7:19,
                                    },
                                ),
                                NonJoint,
                            ),
                            (
                                Delimited(
                                    DelimSpan {
                                        open: src/test/rustdoc/macro-in-closure.rs:7:19: 7:20,
                                        close: src/test/rustdoc/macro-in-closure.rs:7:20: 7:21,
                                    },
                                    Brace,
                                    TokenStream(
                                        [],
                                    ),
                                ),
                                NonJoint,
                            ),
                        ],
                    ),
                ),
                macro_rules: false,
            },
        },
    ],
    non_exported_macro_attrs: [],
    items: {
        HirId {
            owner: DefId(0:1 ~ macro_in_closure[317d]::{{misc}}[0]),
            local_id: 0,
        }: Item {
            ident: #0,
            hir_id: HirId {
                owner: DefId(0:1 ~ macro_in_closure[317d]::{{misc}}[0]),
                local_id: 0,
            },
            attrs: [
                Attribute {
                    kind: Normal(
                        AttrItem {
                            path: Path {
                                span: src/test/rustdoc/macro-in-closure.rs:1:1: 1:1,
                                segments: [
                                    PathSegment {
                                        ident: prelude_import#1,
                                        id: NodeId(4),
                                        args: None,
                                    },
                                ],
                            },
                            args: Empty,
                        },
                    ),
                    id: AttrId(2),
                    style: Outer,
                    span: src/test/rustdoc/macro-in-closure.rs:1:1: 1:1,
                },
            ],
            kind: Use(
                Path {
                    span: src/test/rustdoc/macro-in-closure.rs:1:1: 1:1,
                    res: Err,
                    segments: [
                        PathSegment {
                            ident: {{root}}#1,
                            hir_id: Some(
                                HirId {
                                    owner: DefId(0:1 ~ macro_in_closure[317d]::{{misc}}[0]),
                                    local_id: 1,
                                },
                            ),
                            res: Some(
                                Err,
                            ),
                            args: None,
                            infer_args: false,
                        },
                        PathSegment {
                            ident: std#1,
                            hir_id: Some(
                                HirId {
                                    owner: DefId(0:1 ~ macro_in_closure[317d]::{{misc}}[0]),
                                    local_id: 2,
                                },
                            ),
                            res: Some(
                                Def(
                                    Mod,
                                    DefId(1:0 ~ std[512e]),
                                ),
                            ),
                            args: None,
                            infer_args: false,
                        },
                        PathSegment {
                            ident: prelude#1,
                            hir_id: Some(
                                HirId {
                                    owner: DefId(0:1 ~ macro_in_closure[317d]::{{misc}}[0]),
                                    local_id: 3,
                                },
                            ),
                            res: Some(
                                Def(
                                    Mod,
                                    DefId(1:15 ~ std[512e]::prelude[0]),
                                ),
                            ),
                            args: None,
                            infer_args: false,
                        },
                        PathSegment {
                            ident: v1#1,
                            hir_id: Some(
                                HirId {
                                    owner: DefId(0:1 ~ macro_in_closure[317d]::{{misc}}[0]),
                                    local_id: 4,
                                },
                            ),
                            res: Some(
                                Def(
                                    Mod,
                                    DefId(1:16 ~ std[512e]::prelude[0]::v1[0]),
                                ),
                            ),
                            args: None,
                            infer_args: false,
                        },
                    ],
                },
                Glob,
            ),
            vis: Spanned {
                node: Inherited,
                span: src/test/rustdoc/macro-in-closure.rs:1:1: 1:1,
            },
            span: src/test/rustdoc/macro-in-closure.rs:1:1: 1:1,
        },
        HirId {
            owner: DefId(0:2 ~ macro_in_closure[317d]::std[0]),
            local_id: 0,
        }: Item {
            ident: std#2,
            hir_id: HirId {
                owner: DefId(0:2 ~ macro_in_closure[317d]::std[0]),
                local_id: 0,
            },
            attrs: [
                Attribute {
                    kind: Normal(
                        AttrItem {
                            path: Path {
                                span: src/test/rustdoc/macro-in-closure.rs:1:1: 1:1,
                                segments: [
                                    PathSegment {
                                        ident: macro_use#1,
                                        id: NodeId(10),
                                        args: None,
                                    },
                                ],
                            },
                            args: Empty,
                        },
                    ),
                    id: AttrId(1),
                    style: Outer,
                    span: src/test/rustdoc/macro-in-closure.rs:1:1: 1:1,
                },
            ],
            kind: ExternCrate(
                None,
            ),
            vis: Spanned {
                node: Inherited,
                span: src/test/rustdoc/macro-in-closure.rs:1:1: 1:1,
            },
            span: src/test/rustdoc/macro-in-closure.rs:1:1: 1:1,
        },
        HirId {
            owner: DefId(0:3 ~ macro_in_closure[317d]::main[0]),
            local_id: 0,
        }: Item {
            ident: main#0,
            hir_id: HirId {
                owner: DefId(0:3 ~ macro_in_closure[317d]::main[0]),
                local_id: 0,
            },
            attrs: [],
            kind: Fn(
                FnSig {
                    header: FnHeader {
                        unsafety: Normal,
                        constness: NotConst,
                        asyncness: NotAsync,
                        abi: Rust,
                    },
                    decl: FnDecl {
                        inputs: [],
                        output: DefaultReturn(
                            src/test/rustdoc/macro-in-closure.rs:5:11: 5:11,
                        ),
                        c_variadic: false,
                        implicit_self: None,
                    },
                },
                Generics {
                    params: [],
                    where_clause: WhereClause {
                        predicates: [],
                        span: src/test/rustdoc/macro-in-closure.rs:5:10: 5:10,
                    },
                    span: src/test/rustdoc/macro-in-closure.rs:5:8: 5:8,
                },
                BodyId {
                    hir_id: HirId {
                        owner: DefId(0:3 ~ macro_in_closure[317d]::main[0]),
                        local_id: 6,
                    },
                },
            ),
            vis: Spanned {
                node: Inherited,
                span: src/test/rustdoc/macro-in-closure.rs:5:1: 5:1,
            },
            span: src/test/rustdoc/macro-in-closure.rs:5:1: 9:2,
        },
    },
    trait_items: {},
    impl_items: {},
    bodies: {
        BodyId {
            hir_id: HirId {
                owner: DefId(0:3 ~ macro_in_closure[317d]::main[0]),
                local_id: 2,
            },
        }: Body {
            params: [],
            value: Expr {
                hir_id: HirId {
                    owner: DefId(0:3 ~ macro_in_closure[317d]::main[0]),
                    local_id: 2,
                },
                kind: Block(
                    Block {
                        stmts: [],
                        expr: None,
                        hir_id: HirId {
                            owner: DefId(0:3 ~ macro_in_closure[317d]::main[0]),
                            local_id: 1,
                        },
                        rules: DefaultBlock,
                        span: src/test/rustdoc/macro-in-closure.rs:6:8: 8:6,
                        targeted_by_break: false,
                    },
                    None,
                ),
                attrs: ThinVec(
                    None,
                ),
                span: src/test/rustdoc/macro-in-closure.rs:6:8: 8:6,
            },
            generator_kind: None,
        },
        BodyId {
            hir_id: HirId {
                owner: DefId(0:3 ~ macro_in_closure[317d]::main[0]),
                local_id: 6,
            },
        }: Body {
            params: [],
            value: Expr {
                hir_id: HirId {
                    owner: DefId(0:3 ~ macro_in_closure[317d]::main[0]),
                    local_id: 6,
                },
                kind: Block(
                    Block {
                        stmts: [
                            Stmt {
                                hir_id: HirId {
                                    owner: DefId(0:3 ~ macro_in_closure[317d]::main[0]),
                                    local_id: 4,
                                },
                                kind: Semi(
                                    Expr {
                                        hir_id: HirId {
                                            owner: DefId(0:3 ~ macro_in_closure[317d]::main[0]),
                                            local_id: 3,
                                        },
                                        kind: Closure(
                                            Ref,
                                            FnDecl {
                                                inputs: [],
                                                output: DefaultReturn(
                                                    src/test/rustdoc/macro-in-closure.rs:6:8: 6:8,
                                                ),
                                                c_variadic: false,
                                                implicit_self: None,
                                            },
                                            BodyId {
                                                hir_id: HirId {
                                                    owner: DefId(0:3 ~ macro_in_closure[317d]::main[0]),
                                                    local_id: 2,
                                                },
                                            },
                                            src/test/rustdoc/macro-in-closure.rs:6:5: 6:7,
                                            None,
                                        ),
                                        attrs: ThinVec(
                                            None,
                                        ),
                                        span: src/test/rustdoc/macro-in-closure.rs:6:5: 8:6,
                                    },
                                ),
                                span: src/test/rustdoc/macro-in-closure.rs:6:5: 8:7,
                            },
                        ],
                        expr: None,
                        hir_id: HirId {
                            owner: DefId(0:3 ~ macro_in_closure[317d]::main[0]),
                            local_id: 5,
                        },
                        rules: DefaultBlock,
                        span: src/test/rustdoc/macro-in-closure.rs:5:11: 9:2,
                        targeted_by_break: false,
                    },
                    None,
                ),
                attrs: ThinVec(
                    None,
                ),
                span: src/test/rustdoc/macro-in-closure.rs:5:11: 9:2,
            },
            generator_kind: None,
        },
    },
    trait_impls: {},
    body_ids: [
        BodyId {
            hir_id: HirId {
                owner: DefId(0:3 ~ macro_in_closure[317d]::main[0]),
                local_id: 6,
            },
        },
        BodyId {
            hir_id: HirId {
                owner: DefId(0:3 ~ macro_in_closure[317d]::main[0]),
                local_id: 2,
            },
        },
    ],
    modules: {
        HirId {
            owner: DefId(0:0 ~ macro_in_closure[317d]),
            local_id: 0,
        }: ModuleItems {
            items: {
                HirId {
                    owner: DefId(0:1 ~ macro_in_closure[317d]::{{misc}}[0]),
                    local_id: 0,
                },
                HirId {
                    owner: DefId(0:2 ~ macro_in_closure[317d]::std[0]),
                    local_id: 0,
                },
                HirId {
                    owner: DefId(0:3 ~ macro_in_closure[317d]::main[0]),
                    local_id: 0,
                },
            },
            trait_items: {},
            impl_items: {},
        },
    },
    proc_macros: [],
warning: 1 warning emitted
}
```

cc https://github.com/rust-lang/rust/issues/39412

### PR DESCRIPTION
**Pull request by marmeladema**
Fixes #71820

### CODE REVIEW

File: src/librustc_ast/ast.rs [hunk 1]
```diff
@@ -899,6 +899,33 @@ impl Stmt {
...
+        }
+    }
+
+    pub fn is_semi_expr(&self) -> bool {
+        match self.kind {
+            StmtKind::Semi(_) => true,
+            _ => false,
+        }
+    }
+
```
Comment (lcnr) [inline src/librustc_ast/ast.rs:919]:
```suggestion
```
Seems like you ended up not needing this method afaict

Comment (marmeladema) [reply src/librustc_ast/ast.rs:919]: Yeah i just added it for parity with `StmtKind::Expr`

**Comment by marmeladema** [comment]
cc @petrochenkov

File: src/librustc_interface/util.rs [hunk 2]
```diff
@@ -741,7 +745,11 @@ impl<'a> MutVisitor for ReplaceBodyWithLoop<'a, '_> {
                 for s in b.stmts {
                     let old_blocks = self.nested_blocks.replace(vec![]);
 
-                    stmts.extend(self.flat_map_stmt(s).into_iter().filter(|s| s.is_item()));
+                    stmts.extend(self.flat_map_stmt(s).into_iter().filter(|s| {
+                        s.is_item()
+                            || s.is_expr_with(expr_is_closure)
+                            || s.is_semi_expr_with(expr_is_closure)
```
Comment (ecstatic-morse) [inline src/librustc_interface/util.rs:751]: You should use a single  `fn expr(&self) -> Option<&ast::Expr>` meth `ast::Stmt` instead. I think it's very rare that you would want to call `is_expr_with` without `is_semi_expr_with`.

**Comment by ecstatic-morse** [comment]
What happens if I have code like the following:

```rust
fn foo() {
    let x = |()| {};
}
```

I think the closure still gets removed because it's in a `StmtKind::Local`. Is this what we want? If not, we should map these to `StmtKind::Expr`s instead.

**Comment by marmeladema** [comment]
> What happens if I have code like the following:
> 
> ```rust
> fn foo() {
>     let x = |()| {};
> }
> ```
> 
> I think the closure still gets removed because it's in a `StmtKind::Local`. Is this what we want? If not, we should map these to bare `StmtKind::Expr`s instead.

Yes the closure would probably gets removed, but i guess its the intended behavior of the simplification pass? The problem in my case is a closure with an exported macro inside. I tried various other code layout (useless blocks, nested closures, etc) but was not able to reproduce the bug when applying this change.

I think the real question is whether or not this simplification pass is needed at all? It looks like a weird hack and can imagine a lot of different bugs due to it.

**Comment by marmeladema** [comment]
> What happens if I have code like the following:
> 
> ```rust
> fn foo() {
>     let x = |()| {};
> }
> ```

So i just tested your example with a macro inside, and indeed, you are right, i can reproduce the bug.

> 
> I think the closure still gets removed because it's in a `StmtKind::Local`. Is this what we want? If not, we should map these to bare `StmtKind::Expr`s instead.

Sorry, I am not sure i fully understand why you mean. Could you explain a bit more? Or maybe provide pseudo-code / example?

**Comment by ecstatic-morse** [comment]
My point is that this won't fix #71820 if you have the following, since you're not looking at all places an expressio could appear in a `Stmt`.

```rust
fn foo() {
    let x = |()| {
        #[macro_export] macro_rules! foo {}
    };
}
```

**Comment by ecstatic-morse** [comment]
> I think the real question is whether or not this simplification pass is needed at all? It looks like a weird hack and can imagine a lot of different bugs due to it.

I can't speak to how rustdoc is using this. Perhaps someone else knows? I'd be surprised if leaving in closures caused anything to break, although you should mention why you're doing this in a comment, since this is *very* subtle.

**Comment by marmeladema** [comment]
> Now that I think of it, there are probably more expressions that will introduce a new HirId just as closures do. For example, async blocks probably work like this?

Just tried it, and yes, the bug also happens with async block. Basically, any AST node that can contain a macro and that is going to get removed (ie: not an `Item` node) would trigger the issue.

**Comment by ecstatic-morse** [comment]
So basically we need to make the AST visitor more robust. Specifically, it should only remove AST nodes that contain no nested items. Does this seem feasible? Unfortunately, `ReplaceBodyWithLoop` is getting further away from doing what it promises. I think we will need to port rustdoc to a hypothetical `libsyntax2` to remove it, however.

**Comment by marmeladema** [comment]
I've improved the implem based on your feedbacks @ecstatic-morse ! Thanks a lot.
Should we create a ticket for the proper fix and merge this PR in the mean time?

**Comment by ecstatic-morse** [comment]
I'm good with that. I'll take another look when CI is green.

**Comment by petrochenkov** [comment]
I'm still not sure why this issue happened only with macros but not other items.

**Comment by ecstatic-morse** [comment]
@petrochenkov I have very little intuition for this code, but perhaps it is because we run this pass after macro expansion but before HIR lowering? Maybe inner items that are not macro definitions are never seen by `rustdoc` but macro definitions are collected somewhere that is later processed by `rustdoc`. I would have assumed that `rustdoc` had tests for docs on nested items, however.

**Comment by marmeladema** [comment]
@petrochenkov @ecstatic-morse My understanding is that the bug is triggered because we try to access the `HirId` of a parent `DefId`. This operation does not seem that common(ie: specifically getting the parent `DefId` then getting its `HirId`). Moreover its triggered by macros because during lowering, for the ones that are exported, they are registered in a specific "exported macros" vector which is processed later on for visibility checks.

**Comment by marmeladema** [comment]
So keeping the closures, even in the `StmtKind::Local` case might not the right solution after all because rustdoc now fails in liballoc:
```
let pred = &mut self.pred;
let mut mapped_pred = |k: &T, _v: &mut ()| pred(k);
```
First `let` statement is removed, but second is kept and fails to compile because `pred` does not exist anymore.

Another idea i had, was to monkey patch the parent `DefId` of exported macros during lowering. I don't really know the implications of such an approach.

**Comment by ecstatic-morse** [comment]
>So basically we need to make the AST visitor more robust. Specifically, it should only remove AST nodes that contain no nested items. 

@marmeladema This is what I was hinting at with the comment above. The AST visitor approach will only work if you remove *all* expressions that don't have nested items. Currently, the AST visitor is only handling block expressions, which means you're missing stuff like the single-expression closure above.

Monkey patching `DefId`s does seem more robust than stripping AST nodes, but I also don't know how you would go about it. My position is that `ReplaceBodyWithLoop` is already a hack, so we should keep any fixes internal and avoid changing other parts of the compiler if at all possible.

**Comment by crlf0710** [comment]
@marmeladema Ping from triage. The CI failure needs to be fixed.

**Comment by marmeladema** [comment]
@crlf0710 its broken but i don't know what is the best way to fix yet. Should I move the PR to draft for the time being?

**Comment by ecstatic-morse** [comment]
@marmeladema Do you mind if I take this over?

**Comment by marmeladema** [comment]
Be my guest :+1: For the record, there are some important refactor work going on right now like https://github.com/rust-lang/rust/pull/72284 that might help fixing this.

**Comment by ecstatic-morse** [comment]
>  For the record, there are some important refactor work going on right now like #72284 that might help fixing this.

Sweet. I'll wait for that to land.

**Comment by petrochenkov** [comment]
https://github.com/rust-lang/rust/pull/72284 seems unrelated to this PR.

**Comment by ecstatic-morse** [comment]
@marmeladema Seems like that PR landed. Is this still desirable? Or is there a different approach in the works?

**Comment by marmeladema** [comment]
So #72284 was indeed unrelated, i got confused with another macro def map that exists during lowering .. sorry for the confusion. I haven't time to try any other approach, feel free to pick that up @ecstatic-morse you will probably be more efficient that me at solving that issue :)

**Comment by jyn514** [comment]
Let me know if I can help with this, #71820 is blocking https://github.com/rust-lang/rust/pull/73101.

**Comment by ecstatic-morse** [comment]
Superseded by #73103.

**Pull request closed without merging**